lint/rule.go

```diff
@@ -14,10 +14,172 @@
 
 package lint
 
+import (
+	"fmt"
+	"strings"
+
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+)
+
 // Rule defines a lint rule that checks Google Protobuf APIs.
-type Rule interface {
-	// Info returns metadata about a rule.
-	Info() RuleInfo
-	// Lint performs the linting process.
-	Lint(Request) ([]Problem, error)
+type Rule struct {
+	// Name contains the rule's name.
+	Name RuleName
+
+	// Description is a short description of the rule,
+	// used in documentation and elsewhere.
+	Description string
+
+	// URI is the address where the guideline is documented.
+	// This should be displayed to give API designers more information about
+	// "how to do this right".
+	URI string
+
+	// LintFile is called for files as a whole.
+	LintFile func(*desc.FileDescriptor) []Problem
+
+	// LintMessage is called for any messages in the file.
+	// This includes nested messages, regardless of depth.
+	LintMessage func(*desc.MessageDescriptor) []Problem
+
+	// LintField is called for any fields in the file.
+	// This includes fields within nested messages, regardless of depth.
+	LintField func(*desc.FieldDescriptor) []Problem
+
+	// LintService is called for each service in the file.
+	LintService func(*desc.ServiceDescriptor) []Problem
+
+	// LintMethod is called for each method in the file.
+	LintMethod func(*desc.MethodDescriptor) []Problem
+
+	// LintEnum is called for each enum in the file.
+	LintEnum func(*desc.EnumDescriptor) []Problem
+}
```

Comment:
Have you considered using interfaces instead of a big struct?

```go
type FileLinter interface {
  func LintFile(*desc.FileDescriptor) []Problem
}

type MessageLinter interface {
  func LintMessage(*desc.MessageDescriptor) []Problem
}
```

There would be very little burden on the rule implementer (i.e. just add the funcs that you want to implement), and go's [type switch](https://tour.golang.org/methods/16) makes this implementation really easy and type-safe for the linter core. It's also a bit more idiomatic in Go, and will enable users to implement our desired behavior with their own data structures if needed. To me, this struct looks more like a convenience helper than a library interface, and can be made available as such (i.e. in the linter core we don't actually care about the data structure used to hold state and lint descriptors, but rather the behavior, which is idiomatically described with interfaces).

If you do choose to go this route, I would suggest adding a `noPositional struct{}` argument here to prevent users from initializing this struct with positional arguments.

Reply:
I would be willing to change this to use interfaces in the way that you suggest.

Reply:
Going down this road, this is less clear to me. It might be lack of knowledge on my part, though.

I feel like this is going to end in something like...

```go
type RuleInterface interface {
  Name string
  Description string
  URI string
  Lint func(m *desc.FileDescriptor) []Problem
}

type MessageRuleInterface {
  LintMessage func(m *desc.MessageDescriptor) []Problem
}
```

And then I still need the concrete variant...

```go
struct MessageRule {
  Name string
  Description string
  URI string
  LintMessage func(m *desc.MessageDescriptor) []Problem
}
```

And then the user instantiates a `MessageRule` rather than a `Rule`, otherwise identically. That does not seem like an improvement, it just seems like extra code.

An alternative variant might be to _only_ define the interfaces and then the user defines their own struct...

```go
// Current way...
myRule = Rule{
  Name: "..."
  Description: "..."
  URI: "..."
  LintMessage: func(m *desc.MessageDescriptor) []Problem {
    // implementation here...
  }
}

// New way...
myRule = struct{
  Name string
  Description string
  URI string
  LintMessage: func(m *desc.MessageDescriptor) []Problem
}{
  Name: "..."
  Description: "..."
  URI: "..."
  LintMessage: func(m *desc.MessageDescriptor) []Problem {
    // implementation here...
  }
}
```

...I assume that can not be what you mean, because that seems even more terrible.

What am I missing? :)

Reply:
Sorry, I should have sent a more detailed explanation. Basically what I was thinking was a main `rule` interface that would be embedded in each of the other more specific interfaces, and therefore you could have something like:
```go
type rule interface {
  Info() RuleInfo
}
type fieldLinter interface {
  rule
  LintField(d desc.FieldDescriptor) []Problem
}
type messageLinter interface { ... }
// ...
```
and rules like
```go
type myRule struct{}
func (r myRule) Info() RuleInfo {
  return RuleInfo{
    Name: "...",
    Description: "..."
    URI: "..."
  }
}
func (r myRule) LintField(m *desc.FieldDescriptor) []Problem {
  // implementation here...
}
```
and each rule could choose to implement any number of methods like `LintField` `LintMessage` etc. I'm not sure if it solves the same problem you're after here. It's still trivial to add a helper struct like the one you mentioned above, but by keeping the interfaces as the boundary, people can easily implement their own structures if they need to do things like persist state across calls to LintXxx, and if we want to add a new helper we can easily do so without breaking rules.
 
It's really just an option to consider, because you're making some pretty significant design changes as part of this PR. What I'd really like to see is a drop-in of the new protoreflect library for the old one. It's difficult to judge the impact that the change of library has with the impact of the larger structural changes in the repo, and I think these changes are distracting from the more important library swap. I think I understand why you did it like this (since you're breaking the interface regardless), but I think two smaller scale changes would be easier to review than one larger change. The library swap is pretty cut and dry, but the structural changes have some more important design considerations.

By the way, I started a [go-nuts discussion](http://g/go-nuts/gjPmImTdf94) a while ago about this topic. You might be interested in some of that discussion.

Reply:
> and each rule could choose to implement any number of methods like LintField LintMessage etc.

I am trying to cut down on the amount of boilerplate in rule-writing and rule-reading, and this re-introduces a significant amount of it. I fail to see why having an `Info()` method as a level of indirection is valuable, for example. When I asked @alin04 about it, she felt similarly.

I guess if I want to do this, I want to understand why it is _better_, because it really just feels like a more complicated way to accomplish the same thing.

> I think I understand why you did it like this (since you're breaking the interface regardless), but I think two smaller scale changes would be easier to review than one larger change. The library swap is pretty cut and dry, but the structural changes have some more important design considerations.

Right, that requires two separate migrations before we can start writing rules. I am trying to address any interface-breaking concerns so we can "get this show on the road", so to speak.

I _can_ break this up into two PRs, but it would be a significant amount of work, and I am disinclined unless we need to address a design concern.

Reply:
I talked to @noahdietz and he has somewhat convinced me.

I think implementing interfaces here is reasonable, but it is _really_ important to me that we do not have a lot of extra boilerplate code when defining most rules. I think I can make that happen though.

Reply:
`RuleInfo` is just a way to store structured data that doesn't have any behavior attached to it. It's essentially the same thing as a proto `message` – a declaration of data that we expect to be associated with rules. `Rule` is closer to a `service` definition – a declaration of behavior we expect without any ties to implementation. Adding `Info()` to the rule interface enables us to make sure that every rule implementation provides us with the data we care about. A side effect is that you can easily reuse the same set of info for many rules. For example, you can use a factory method or shared global struct across multiple rules and keep their shared info in one place (e.g. for an AIP that has a bunch of rules). The benefit of using an interface for rules is that it doesn't bind us to any implementation. That's why you can easily create the code you have here as a helper and satisfy the interface with it, and if I want to implement a rule using the code snippet in my previous comment, I can do that as well. Everyone's happy.

By my count, it's 7 lines of boilerplate versus 4. Is the enforced reduction of 3 lines of boilerplate enough to justify binding rule developers to a specific implementation and restructuring a bunch of existing code? I think the primary point that i'm trying (maybe poorly) to communicate here is that _interfaces are cheap_. Especially if they're unexported like the example I showed above. You can delete, rename, or restructure them as needed, because they're package private and no one depends on them but the code you already own. It's basically as cheap as python's duck typing, but you get free compile time checks with it.

Because you've already offered to refactor all the google3 code that's tied to the existing interface, I won't block this. And of course I'll be happy to help with any questions you have when you do the first sync.

Reply:
Just noticed you're other message. Yes, I think you're rightfully concerned about boilerplate. Helper libraries / method can help a lot here. Glad @noahdietz is more convincing than me :)